### PR TITLE
Add more robust e2eTestPrompt.md namespace to AssemblyName mapping. Filter prompts that need investigation

### PR DESCRIPTION
**Pull request by conniey**
## What does this PR do?

Improves VallyEvaluator namespace selection when an MCP tool assembly name does not directly match its prompt namespace.  For example, `Azure.Mcp.Tools.AzureIsv`'s namespace is `datadog`, which cannot be easily inferred when parsing e2eTestPrompts.md

Adds Interaction.InvestigationRequired value for prompts that require change (either with tool description or the prompt itself.)

The evaluator now:
- Maps changed tool projects from `build_info.json` to their registered MCP tool area namespaces.
- Selects the server artifact matching the current runtime and supports choosing a server with `--serverName`.
- Adds a VallyEvaluator README covering setup, generation options, build-info mode, and local execution.

## Pre-merge Checklist
- [ ] Required for All PRs
    - [ ] **Read [contribution guidelines](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md)**
    - [ ] PR title clearly describes the change
    - [ ] Commit history is clean with descriptive messages ([cleanup guide](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md))
    - [ ] Added comprehensive tests for new/modified functionality
    - [ ] Created a changelog entry if the change falls among the following: new feature, bug fix, UI/UX update, breaking change, or updated dependencies. Follow [the changelog entry guide](https://github.com/microsoft/mcp/blob/main/docs/changelog-entries.md)
- [ ] For MCP tool changes:
    - [ ] **One tool per PR**: This PR adds or modifies only one MCP tool for faster review cycles
    - [ ] Updated `servers/Azure.Mcp.Server/README.md` and/or `servers/Fabric.Mcp.Server/README.md` documentation
    - [ ] Validate `README.md` changes running the script `./eng/scripts/Process-PackageReadMe.ps1`. See [Package README](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md#package-readme)
    - [ ] For new or modified tool descriptions, ran [`ToolDescriptionEvaluator`](https://github.com/microsoft/mcp/blob/main/eng/tools/ToolDescriptionEvaluator/Quickstart.md) and obtained a score of `0.4` or more and a top 3 ranking for all related test prompts
    - [ ] For tools with new names, including new tools or renamed tools, update [`consolidated-tools.json`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/src/Resources/consolidated-tools.json)
    - [ ] For **renamed** tools, follow the [Tool Rename Checklist](https://github.com/microsoft/mcp/blob/main/docs/tool-rename-checklist.md) and tag the PR with the `breaking-change` label
    - [ ] For new tools associated with Azure services or publicly available tools/APIs/products, add URL to documentation in the PR description
- [ ] Extra steps for **Azure MCP Server** tool changes:
    - [ ] Updated command list in [`servers/Azure.Mcp.Server/docs/azmcp-commands.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/azmcp-commands.md)
    - [ ] Ran `./eng/scripts/Update-AzCommandsMetadata.ps1` to update tool metadata in [`azmcp-commands.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/azmcp-commands.md) (required for CI)
    - [ ] Updated test prompts in [`servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md)
    - [ ] 👉 For Community (non-Microsoft team member) PRs:
        - [ ] **Security review**: Reviewed code for security vulnerabilities, malicious code, or suspicious activities before running tests (`crypto mining, spam, data exfiltration, etc.`)
        - [ ] **Manual tests run**: added comment `/azp run mcp - pullrequest - live` to run *Live Test Pipeline*

### CODE REVIEW

File: eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    internal void ProbeSetupNames(string targetPath)
+    {
+        targetPath = Path.GetFullPath(targetPath);
+
+        Console.WriteLine($"Probing {targetPath} for classes...");
+
+        // Scan the folder directly rather than reading the target's AssemblyReferences.
+        // A folder scan can't miss a Setup assembly that the target references but never
+        // uses a type from (the compiler can trim such references out of the metadata).
+        foreach (string dll in Directory.EnumerateFiles(targetPath, "*.dll"))
```
Comment (alzimmermsft) [inline eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs:43]: This might have some edge cases as the Core library DLLs will exist in the Tool library folders. Or, is the docstring a bit misleading as this will only be scanning `/server` folders?

File: eng/tools/McpToolEvaluator.Core/src/McpServerInformation.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                // managed assemblies. Not a .NET assembly, so skip it.
+            }
+        }
+    }
+
+    /// <summary>
+    /// Opens a single assembly's metadata and appends a ToolArea for every "*Setup" type
+    /// that has a Name property returning a string literal.
+    /// </summary>
+    internal static List<ToolArea> ProbeAssembly(string path)
```
Comment (alzimmermsft) [inline eng/tools/McpToolEvaluator.Core/src/McpServerInformation.cs]: Would it help if we added a concept like `NamespaceMetadata`, that is used on `<Area>Setup` types? Similar to what we did with `CommandMetadata` for tools.

File: eng/tools/VallyEvaluator/src/Program.cs [hunk 3]
```diff
@@ -17,24 +19,46 @@ public static async Task<int> Main(string[] args)
         var runConfig = new RunConfiguration();
         configuration.Bind(runConfig);
 
+        if (string.IsNullOrEmpty(runConfig.ServerName))
+        {
+            Console.WriteLine("No MCP server specified. Using default: Azure.Mcp.Server");
+            runConfig.ServerName = "Azure.Mcp.Server";
```
Comment (alzimmermsft) [inline eng/tools/VallyEvaluator/src/Program.cs]: Did we think about throwing an error here instead? Basically, require server name to be passed so the caller explicitly knows what it is evaluating to prevent cases where this "works" and looks good without deeper review but isn't testing what is expected.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -252,11 +252,11 @@ Real product code under unit testing must be passed `Xunit.TestContext.Current.C
 
 ### End-to-end Tests
 
-End-to-end tests are performed manually. Command authors must thoroughly test each command to ensure correct tool invocation and results. At least one prompt per tool is required and should be added to `/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`.
+End-to-end tests are performed manually. Command authors must thoroughly test each command to ensure correct tool invocation and results. At least one prompt per tool is required and should be added to [`/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`](https://github.com/microsoft/mcp/blob/main/servers/Azure.Mcp.Server/docs/e2eTestPrompts.md).
```
Comment (alzimmermsft) [inline CONTRIBUTING.md]: This is the root-level `CONTRIBUTING.md`, mind if we make the verbiage more generic?


```suggestion
End-to-end tests are performed manually. Command authors must thoroughly test each command to ensure correct tool invocation and results. At least one prompt per tool is required and should be added to the server's `/server/<servername>/docs/e2eTestPrompts.md` file.
```

File: eng/tools/VallyEvaluator/README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+- The .NET SDK selected by the repository's `global.json`
+- [PowerShell 7](https://learn.microsoft.com/powershell/scripting/install/installing-powershell) to use the repository scripts
+- [Vally](https://microsoft.github.io/vally/get-started/install/) to run evaluations
+- [GitHub Copilot CLI](https://docs.github.com/en/copilot/how-tos/copilot-sdk/setup/local-cli), authenticated with `copilot login`
+
+CI installs the Vally CLI with:
+
+```powershell
+npm install --global @microsoft/vally-cli@0.13.0
```
Comment (alzimmermsft) [inline eng/tools/VallyEvaluator/README.md:17]: Do we want to mention the specific version of `vally-cli` being installed? This will get out of date very soon and this kind of documentation generally is forgotten to be updated.

File: eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            try
+            {
+                foreach (var toolArea in ProbeAssembly(dll))
+                {
+                    AssemblyNameToToolsMap[toolArea.AssemblyName] = toolArea;
+                }
+            }
+            catch (BadImageFormatException)
+            {
```
Comment (g2vinay) [inline eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs:58]: Could this map each assembly to a collection of namespaces, or reject assemblies with multiple discovered areas? The current assignment silently retains only the last namespace.

File: eng/tools/VallyEvaluator/src/Program.cs [hunk 7]
```diff
@@ -115,7 +179,7 @@ private static async Task CreateEvalsAsync(string repoRoot, RunConfiguration con
 
         if (buildInfo != null)
         {
-            var testToolNamespaces = GetTestToolNamespaces(buildInfo, promptDatastore);
+            var testToolNamespaces = GetTestToolNamespaces(buildInfo, promptDatastore, mcpServerInformation);
             if (testToolNamespaces.Count == 0)
             {
                 Console.WriteLine("No valid namespaces found in build info. Exiting.");
```
Comment (g2vinay) [inline eng/tools/VallyEvaluator/src/Program.cs:185]: Could we fail the run when build info contains changed tools but none can be mapped? Returning normally makes `Main` exit 0, so a mapping regression could appear successful with zero generated tests.

File: eng/tools/VallyEvaluator/src/Program.cs [hunk 8]
```diff
@@ -17,24 +19,46 @@ public static async Task<int> Main(string[] args)
...
+            Console.WriteLine("No working directory specified. Using default: " + workDir);
         }
 
         if (!Directory.Exists(runConfig.WorkingDirectory))
         {
             Directory.CreateDirectory(runConfig.WorkingDirectory);
         }
 
-        BuildInfo? buildInfo = null;
+        if (string.IsNullOrEmpty(runConfig.PromptFilePath))
```
Comment (g2vinay) [inline eng/tools/VallyEvaluator/src/Program.cs]: Could the default prompt file be derived from `--serverName`, or could `--promptFile` be required for non-Azure servers? Currently, selecting `Fabric.Mcp.Server` still uses Azure prompts by default.

File: eng/tools/VallyEvaluator/tests/VallyUtilitiesTests.cs
Comment (g2vinay) [inline eng/tools/VallyEvaluator/tests/VallyUtilitiesTests.cs:1]: Could we add focused tests for metadata probing, nonstandard mappings such as `AzureIsv -> datadog`, server/platform selection, zero-match behavior, and stale generated-file cleanup?

File: eng/tools/VallyEvaluator/src/Program.cs [hunk 9]
```diff
@@ -147,20 +211,27 @@ private static async Task CreateEvalsAsync(string repoRoot, RunConfiguration con
...
+
+            if (!noInteractionPrompts.Any())
+            {
+                Console.WriteLine($"- All prompts require interaction for namespace: {ns}");
+                continue;
+            }
+
             var outputDirectory = Path.Combine(vallyEvalDirectory, ns);
             Directory.CreateDirectory(outputDirectory);
             var outputFile = Path.Combine(outputDirectory, "eval.yaml");
```
Comment (g2vinay) [inline eng/tools/VallyEvaluator/src/Program.cs:232]: Could we remove an existing namespace's `eval.yaml` when all its prompts become interactive? Otherwise, `Invoke-VallyEvalTests.ps1` may still execute stale prompts this change intends to exclude.

**Review by g2vinay — CHANGES_REQUESTED** [review]
The assembly-to-registered-area mapping is a useful improvement, but there are a few correctness gaps that can silently skip intended coverage or run stale evaluations. I left focused questions on those paths and the missing behavioral tests.

File: eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs (hunk 1 above)
Comment (conniey) [reply eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs:43]: This scans the directory that the artifact azmcp.exe was built into. All the binaries will live next to it.

File: eng/tools/VallyEvaluator/src/Program.cs (hunk 3 above)
Comment (conniey) [reply eng/tools/VallyEvaluator/src/Program.cs]: I was following the current convention of other Analyze-Code scripts where it assumes that it is Azure.Mcp.Server if nothing is specified.

File: eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs (hunk 6 above)
Comment (conniey) [reply eng/tools/McpToolEvaluator.Core/src/McpServerMetadata.cs:58]: Good catch!

File: eng/tools/VallyEvaluator/src/Program.cs (hunk 7 above)
Comment (conniey) [reply eng/tools/VallyEvaluator/src/Program.cs:185]: build_info.json doesn't contain the diffs themselves but files that were changed. There is another in analyze-code, that we could perform that check into: https://github.com/microsoft/mcp/blob/main/eng/scripts/Test-ToolSelectionPrompts.ps1#L309C1-L310C1

File: eng/tools/VallyEvaluator/src/Program.cs (hunk 8 above)
Comment (conniey) [reply eng/tools/VallyEvaluator/src/Program.cs]: Mm yeah. that would remove the need for the --promptFile parameter. Let me update that.

File: eng/tools/VallyEvaluator/src/Program.cs (hunk 9 above)
Comment (conniey) [reply eng/tools/VallyEvaluator/src/Program.cs:232]: Can you explain this some more? It always regenerates the prompts from e2etestprompts.md into the .work folder, so there should never be stale prompts.

File: eng/tools/VallyEvaluator/src/Program.cs (hunk 3 above)
Comment (conniey) [reply eng/tools/VallyEvaluator/src/Program.cs]: Added this as a required parameter now

File: eng/tools/VallyEvaluator/src/Program.cs (hunk 7 above)
Comment (g2vinay) [reply eng/tools/VallyEvaluator/src/Program.cs:185]: Since build_info.json lists changed files, if GetToolNamespacesFromBuildInfo can't map a changed tool path, it drops it to Console.Error.
So in the edge case when all changed tools fail to map, won't (Count == 0 -> return -> Main exit 0) make a mapping regression look green with zero tests? 
Should we fail non-zero when a changed path under tools yields no namespace?

Comment (conniey) [reply eng/tools/VallyEvaluator/src/Program.cs:185]: https://github.com/microsoft/mcp/issues/3309

**Review by alzimmermsft — APPROVED** [review]

**Review by g2vinay — APPROVED** [review]
